distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/MachineDependentHeap.java

```diff
@@ -0,0 +1,247 @@
+    private static final int DEFAULT_HEAP_SIZE_MB = 1024;
+    private static final String ELASTICSEARCH_YML = "elasticsearch.yml";
+
+    private final SystemMemoryInfo systemMemoryInfo;
+
+    public MachineDependentHeap(SystemMemoryInfo systemMemoryInfo) {
+        this.systemMemoryInfo = systemMemoryInfo;
+    }
+
+    /**
+     * Calculate heap options.
+     *
+     * @param configDir path to config directory
+     * @param userDefinedJvmOptions JVM arguments provided by the user
+     * @return final heap options, or an empty collection if user provided heap options are to be used
+     * @throws IOException if unable to load elasticsearch.yml
+     */
+    public List<String> determineHeapSettings(Path configDir, List<String> userDefinedJvmOptions) throws IOException {
+        if (userDefinedJvmOptions.stream().anyMatch(s -> s.startsWith("-Xms") || s.startsWith("-Xmx"))) {
+            // User has explicitly set memory settings so we use those
+            return Collections.emptyList();
+        }
+
+        Path config = configDir.resolve(ELASTICSEARCH_YML);
+        try (InputStream in = Files.newInputStream(config)) {
+            return determineHeapSettings(in);
+        }
+    }
+
+    List<String> determineHeapSettings(InputStream config) {
+        MachineNodeRole nodeRole = NodeRoleParser.parse(config);
+
+        try {
+            long availableSystemMemory = systemMemoryInfo.availableSystemMemory();
+            return options(nodeRole.heap(availableSystemMemory));
+        } catch (SystemMemoryInfo.SystemMemoryInfoException e) {
+            // If unable to determine system memory (ex: incompatible jdk version) fallback to defaults
+            return options(DEFAULT_HEAP_SIZE_MB);
+        }
+    }
+
+    private static List<String> options(int heapSize) {
+        return List.of("-Xms" + heapSize + "m", "-Xmx" + heapSize + "m");
+    }
+
+    /**
+     * Parses role information from elasticsearch.yml and determines machine node role.
+     */
+    static class NodeRoleParser {
+        private static final Set<String> LEGACY_ROLE_SETTINGS = Set.of(
+            "node.master",
+            "node.ingest",
+            "node.data",
+            "node.voting_only",
+            "node.ml",
+            "node.transform",
+            "node.remote_cluster_client"
+        );
+
+        @SuppressWarnings("unchecked")
+        public static MachineNodeRole parse(InputStream config) {
+            Yaml yaml = new Yaml();
+            Map<String, Object> root;
+            try {
+                root = yaml.load(config);
+            } catch (ClassCastException ex) {
+                // Strangely formatted config, so just return defaults and let startup settings validation catch the problem
+                return MachineNodeRole.UNKNOWN;
+            } catch (YAMLException ex) {
+                throw new IllegalStateException("Unable to parse elasticsearch.yml:", ex);
```

Comment:
I wonder if this should contain the full path? Or if we should return UNKNOWN and let startup fail us here too?

Reply:
I like returning UNKOWN and letting Elasticsearch handle the misconfiguration.